File: litho-it/src/test/java/com/facebook/litho/VisibilityEventsWithVisibilityExtensionTest.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.litho;

import static com.facebook.litho.SizeSpec.EXACTLY;
import static com.facebook.litho.SizeSpec.makeSizeSpec;
import static com.facebook.litho.testing.TestViewComponent.create;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import android.graphics.Rect;
import android.widget.FrameLayout;
import com.facebook.litho.config.ComponentsConfiguration;
import com.facebook.litho.testing.LegacyLithoViewRule;
import com.facebook.litho.testing.TestComponent;
import com.facebook.litho.testing.Whitebox;
import com.facebook.litho.testing.testrunner.LithoTestRunner;
import com.facebook.litho.widget.Text;
import com.facebook.rendercore.MountDelegateTarget;
import com.facebook.rendercore.Reducer;
import com.facebook.rendercore.RenderTree;
import com.facebook.rendercore.RenderTreeNode;
import com.facebook.rendercore.extensions.ExtensionState;
import com.facebook.rendercore.visibility.VisibilityMountExtension;
import com.facebook.yoga.YogaEdge;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

@RunWith(LithoTestRunner.class)
public class VisibilityEventsWithVisibilityExtensionTest {
  private static final int LEFT = 0;
  private static final int RIGHT = 10;

  private ComponentContext mContext;
  private LithoView mLithoView;
  private FrameLayout mParent;

  public final @Rule LegacyLithoViewRule mLegacyLithoViewRule = new LegacyLithoViewRule();

  @Before
  public void setup() {
    mContext = mLegacyLithoViewRule.getContext();
    mLithoView = new LithoView(mContext);
    mLegacyLithoViewRule.useLithoView(mLithoView);

    mParent = new FrameLayout(mContext.getAndroidContext());
    mParent.setLeft(0);
    mParent.setTop(0);
    mParent.setRight(10);
    mParent.setBottom(10);
    mParent.addView(mLithoView);
  }

  @Test
  public void visibilityExtensionInput_dirtyMountWithoutVisibilityProcessing_setInput() {
    final TestComponent content = create(mContext).build();
    final EventHandler<VisibleEvent> visibleEventHandler = new EventHandler<>(content, 2);

    final Component root =
        Column.create(mContext)
            .child(
                Wrapper.create(mContext)
                    .delegate(content)
                    .visibleHandler(visibleEventHandler)
                    .widthPx(10)
                    .heightPx(5)
                    .marginPx(YogaEdge.TOP, 5))
            .build();

    mLegacyLithoViewRule
        .setRoot(root)
        .attachToWindow()
        .setSizeSpecs(makeSizeSpec(10, EXACTLY), makeSizeSpec(5, EXACTLY))
        .measure();

    final LayoutState layoutState = mock(LayoutState.class);
    final RenderTree renderTree = mock(RenderTree.class);
    final RenderTreeNode rootNode = mock(RenderTreeNode.class);
    when(layoutState.toRenderTree()).thenReturn(renderTree);
    when(renderTree.getRenderTreeData()).thenReturn(layoutState);
    when(renderTree.getRenderTreeNodeAtIndex(0)).thenReturn(rootNode);
    when(rootNode.getRenderUnit()).thenReturn(Reducer.sRootHostRenderUnit);

    mLegacyLithoViewRule.getLithoView().setMountStateDirty();

    VisibilityMountExtension visibilityExtension = spy(VisibilityMountExtension.getInstance());

    useVisibilityOutputsExtension(mLegacyLithoViewRule.getLithoView(), visibilityExtension);

    final Rect rect = new Rect(0, 0, RIGHT, 10);

    mLegacyLithoViewRule.getLithoView().mount(layoutState, new Rect(0, 0, RIGHT, 10), false);

    verify(visibilityExtension).beforeMount(any(ExtensionState.class), eq(layoutState), eq(rect));
  }

  @Test
  public void visibilityExtensionOnUnmountAllItems_shouldUnmount() {
    final Component content = Text.create(mContext).text("hello world").build();
    final EventHandler<VisibleEvent> visibleEventHandler = new EventHandler<>(content, 2);

    final Component root =
        Column.create(mContext)
            .child(Wrapper.create(mContext).delegate(content).visibleHandler(visibleEventHandler))
            .build();

    mLegacyLithoViewRule.setRoot(root).attachToWindow().measure().layout();

    final LayoutState layoutState = mock(LayoutState.class);
    final RenderTree renderTree = mock(RenderTree.class);
    final RenderTreeNode rootNode = mock(RenderTreeNode.class);
    when(layoutState.toRenderTree()).thenReturn(renderTree);
    when(renderTree.getRenderTreeData()).thenReturn(layoutState);
    when(renderTree.getRenderTreeNodeAtIndex(0)).thenReturn(rootNode);
    when(rootNode.getRenderUnit()).thenReturn(Reducer.sRootHostRenderUnit);

    mLegacyLithoViewRule.getLithoView().setMountStateDirty();

    VisibilityMountExtension visibilityExtension = spy(VisibilityMountExtension.getInstance());

    useVisibilityOutputsExtension(mLegacyLithoViewRule.getLithoView(), visibilityExtension);

    mLegacyLithoViewRule.getLithoView().unmountAllItems();

    verify(visibilityExtension).onUnbind(any(ExtensionState.class));
    verify(visibilityExtension).onUnmount(any(ExtensionState.class));
  }

  private void useVisibilityOutputsExtension(
      LithoView lithoView, VisibilityMountExtension visibilityOutputsExtension) {
    if (ComponentsConfiguration.delegateToRenderCoreMount) {
      LithoHostListenerCoordinator lithoHostListenerCoordinator =
          Whitebox.getInternalState(lithoView, "mLithoHostListenerCoordinator");
      lithoHostListenerCoordinator.useVisibilityExtension(visibilityOutputsExtension);
    } else {
      final MountDelegateTarget mountState = lithoView.getMountDelegateTarget();
      Whitebox.setInternalState(mountState, "mVisibilityExtension", visibilityOutputsExtension);
    }
  }
}

File: litho-it/src/test/java/com/facebook/litho/VisibilityEventsWithVisibilityExtensionTest.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.litho

import android.graphics.Rect
import android.widget.FrameLayout
import com.facebook.litho.Column.Companion.create
import com.facebook.litho.ComponentsSystrace.systrace
import com.facebook.litho.testing.LegacyLithoViewRule
import com.facebook.litho.testing.TestComponent
import com.facebook.litho.testing.TestViewComponent
import com.facebook.litho.testing.Whitebox
import com.facebook.litho.testing.exactly
import com.facebook.litho.testing.testrunner.LithoTestRunner
import com.facebook.litho.widget.Text
import com.facebook.rendercore.Reducer
import com.facebook.rendercore.RenderTree
import com.facebook.rendercore.RenderTreeNode
import com.facebook.rendercore.visibility.VisibilityMountExtension
import com.facebook.yoga.YogaEdge
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.kotlin.any
import org.mockito.kotlin.eq
import org.mockito.kotlin.mock
import org.mockito.kotlin.spy
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever

@RunWith(LithoTestRunner::class)
class VisibilityEventsWithVisibilityExtensionTest {

  private val right = 10

  private lateinit var context: ComponentContext
  private lateinit var lithoView: LithoView
  private lateinit var parent: FrameLayout

  @JvmField @Rule val legacyLithoViewRule = LegacyLithoViewRule()

  @Before
  fun setup() {
    context = legacyLithoViewRule.context
    lithoView = LithoView(context)
    legacyLithoViewRule.useLithoView(lithoView)
    parent =
        FrameLayout(context.androidContext).apply {
          left = 0
          top = 0
          right = 10
          bottom = 10
          addView(lithoView)
        }
  }

  @Test
  fun visibilityExtensionInput_dirtyMountWithoutVisibilityProcessing_setInput() {
    val content: TestComponent = TestViewComponent.create(context).build()
    val visibleEventHandler = EventHandler<VisibleEvent>(content, 2)
    val root: Component =
        create(context)
            .child(
                Wrapper.create(context)
                    .delegate(content)
                    .visibleHandler(visibleEventHandler)
                    .widthPx(10)
                    .heightPx(5)
                    .marginPx(YogaEdge.TOP, 5))
            .build()
    legacyLithoViewRule
        .setRoot(root)
        .attachToWindow()
        .setSizeSpecs(exactly(10), exactly(5))
        .measure()
    val layoutState: LayoutState = mock()
    val renderTree: RenderTree = mock()
    val rootNode: RenderTreeNode = mock()
    whenever(layoutState.toRenderTree()).thenReturn(renderTree)
    whenever(layoutState.tracer).thenReturn(systrace)
    whenever(renderTree.getRenderTreeNodeAtIndex(0)).thenReturn(rootNode)
    whenever(rootNode.renderUnit).thenReturn(Reducer.sRootHostRenderUnit)
    legacyLithoViewRule.lithoView.setMountStateDirty()
    val visibilityExtension: VisibilityMountExtension<LayoutState> =
        spy(VisibilityMountExtension.getInstance() as VisibilityMountExtension<LayoutState>)
    useVisibilityOutputsExtension(legacyLithoViewRule.lithoView, visibilityExtension)
    val rect = Rect(0, 0, right, 10)
    legacyLithoViewRule.lithoView.mount(layoutState, Rect(0, 0, right, 10), false)
    verify(visibilityExtension).beforeMount(any(), eq(layoutState), eq(rect))
  }

  @Test
  fun visibilityExtensionOnUnmountAllItems_shouldUnmount() {
    val content: SpecGeneratedComponent = Text.create(context).text("hello world").build()
    val visibleEventHandler = EventHandler<VisibleEvent>(content, 2)
    val root: Component =
        create(context)
            .child(Wrapper.create(context).delegate(content).visibleHandler(visibleEventHandler))
            .build()
    legacyLithoViewRule.setRoot(root).attachToWindow().measure().layout()
    val layoutState: LayoutState = mock()
    val renderTree: RenderTree = mock()
    val rootNode: RenderTreeNode = mock()
    whenever(layoutState.toRenderTree()).thenReturn(renderTree)
    whenever(renderTree.getRenderTreeNodeAtIndex(0)).thenReturn(rootNode)
    whenever(rootNode.renderUnit).thenReturn(Reducer.sRootHostRenderUnit)
    legacyLithoViewRule.lithoView.setMountStateDirty()
    val visibilityExtension: VisibilityMountExtension<*> =
        spy(VisibilityMountExtension.getInstance())
    useVisibilityOutputsExtension(legacyLithoViewRule.lithoView, visibilityExtension)
    legacyLithoViewRule.lithoView.unmountAllItems()
    verify(visibilityExtension).onUnbind(any())
    verify(visibilityExtension).onUnmount(any())
  }

  private fun useVisibilityOutputsExtension(
      lithoView: LithoView,
      visibilityOutputsExtension: VisibilityMountExtension<*>
  ) {
    val lithoHostListenerCoordinator =
        Whitebox.getInternalState<LithoHostListenerCoordinator>(
            lithoView, "mLithoHostListenerCoordinator")
    lithoHostListenerCoordinator.useVisibilityExtension(visibilityOutputsExtension)
  }
}
